server/src/main/java/com/linecorp/centraldogma/server/internal/storage/MigratingMetaToDogmaRepositoryService.java

```diff
@@ -0,0 +1,171 @@
+            MigratingMetaToDogmaRepositoryService.class);
+
+    // Will be stored in dogma/dogma
+    public static final String META_TO_DOGMA_MIGRATION_JOB = "/meta-to-dogma-migration-job.json";
+
+    // Will be stored in {project}/dogma
+    public static final String META_TO_DOGMA_MIGRATED = "/meta-to-dogma-migrated.json";
+
+    private final ProjectManager projectManager;
+    private final CommandExecutor commandExecutor;
+    private final MetadataService metadataService;
+
+    MigratingMetaToDogmaRepositoryService(ProjectManager projectManager, CommandExecutor commandExecutor,
+                                          InternalProjectInitializer internalProjectInitializer) {
+        this.projectManager = projectManager;
+        this.commandExecutor = commandExecutor;
+        metadataService = new MetadataService(projectManager, commandExecutor, internalProjectInitializer);
+    }
+
+    boolean hasMigrationLog() throws ExecutionException, InterruptedException {
+        final Project internalProj = projectManager.get(INTERNAL_PROJECT_DOGMA);
+        final Repository repository = internalProj.repos().get(REPO_DOGMA);
+        final Entry<JsonNode> entry =
+                repository.getOrNull(Revision.HEAD, Query.ofJson(META_TO_DOGMA_MIGRATION_JOB)).get();
+        return entry != null;
+    }
+
+    void migrate() throws Exception {
+        logger.info("Starting to migrate meta repository to dogma repository of all projects ...");
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        int numMigratedProjects = 0;
+
+        for (Project project : projectManager.list().values()) {
+            final String projectName = project.name();
+            if (INTERNAL_PROJECT_DOGMA.equals(projectName)) {
+                continue;
+            }
+
+            if (REPO_DOGMA.equals(project.metaRepo().name())) {
+                logger.debug("The project '{}' has already been migrated.", projectName);
+                continue;
+            }
+            logger.info("Migrating meta repository to dogma repository in the project: {} ...", projectName);
+
+            metadataService.updateRepositoryStatus(
+                    Author.SYSTEM, projectName, REPO_DOGMA, RepositoryStatus.READ_ONLY).join();
+            // Wait for the repository status to be updated.
+            Thread.sleep(3000L);
+            logger.info("Dogma repository in the project '{}' is set to read-only.", projectName);
+            try {
+                migrateMetaRepository(project);
+                numMigratedProjects++;
+                logger.info("Meta repository in the project '{}' is migrated to dogma repository.",
+                            projectName);
+            } catch (Throwable t) {
+                logger.warn("Failed to migrate meta repository. project: {}", projectName, t);
+                // Do not continue the migration if the migration fails.
+                return;
```

Comment:
If a migration fails, would it be safer to just set `RepositoryStatus` back to active?